diff --git a/source/d3d8to9.cpp b/source/d3d8to9.cpp
index d2082aa..fee6126 100644
--- a/source/d3d8to9.cpp
+++ b/source/d3d8to9.cpp
@@ -2,7 +2,7 @@
  * Copyright (C) 2015 Patrick Mours. All rights reserved.
  * License: https://github.com/crosire/d3d8to9#license
  */
-
+#include "pch.h"
 #include "d3dx9.hpp"
 #include "d3d8to9.hpp"
 
@@ -15,7 +15,7 @@ PFN_D3DXLoadSurfaceFromSurface D3DXLoadSurfaceFromSurface = nullptr;
 std::ofstream LOG;
 #endif
 
-extern "C" IDirect3D8 *WINAPI Direct3DCreate8(UINT SDKVersion)
+extern "C" IDirect3D8 * WINAPI Direct3DCreate8(UINT SDKVersion)
 {
 #ifndef D3D8TO9NOLOG
 	static bool LogMessageFlag = true;
diff --git a/source/d3d8to9.hpp b/source/d3d8to9.hpp
index 9a7b5df..1e56463 100644
--- a/source/d3d8to9.hpp
+++ b/source/d3d8to9.hpp
@@ -9,6 +9,8 @@
 #include "d3d8.hpp"
 #include "interface_query.hpp"
 
+extern "C" IDirect3D8 * WINAPI Direct3DCreate8(UINT SDKVersion);
+
 class Direct3D8 : public IDirect3D8
 {
 	Direct3D8(const Direct3D8 &) = delete;
diff --git a/source/d3d8to9_base.cpp b/source/d3d8to9_base.cpp
index 5aca4af..d215673 100644
--- a/source/d3d8to9_base.cpp
+++ b/source/d3d8to9_base.cpp
@@ -2,7 +2,7 @@
  * Copyright (C) 2015 Patrick Mours. All rights reserved.
  * License: https://github.com/crosire/d3d8to9#license
  */
-
+#include "pch.h"
 #include "d3d8to9.hpp"
 
 static const D3DFORMAT AdapterFormats[] = {
diff --git a/source/d3d8to9_device.cpp b/source/d3d8to9_device.cpp
index b3a2202..071e1d8 100644
--- a/source/d3d8to9_device.cpp
+++ b/source/d3d8to9_device.cpp
@@ -2,12 +2,14 @@
  * Copyright (C) 2015 Patrick Mours. All rights reserved.
  * License: https://github.com/crosire/d3d8to9#license
  */
-
+#include "pch.h"
 #include "d3dx9.hpp"
 #include "d3d8to9.hpp"
 #include <regex>
 #include <assert.h>
 
+#include "App.h"
+
 struct VertexShaderInfo
 {
 	IDirect3DVertexShader9 *Shader;
@@ -17,6 +19,9 @@ struct VertexShaderInfo
 Direct3DDevice8::Direct3DDevice8(Direct3D8 *d3d, IDirect3DDevice9 *ProxyInterface, BOOL EnableZBufferDiscarding) :
 	D3D(d3d), ProxyInterface(ProxyInterface), ZBufferDiscarding(EnableZBufferDiscarding)
 {
+	if (App::Instance().CreateDeviceCallback)
+		App::Instance().CreateDeviceCallback(ProxyInterface);
+
 	ProxyAddressLookupTable = new AddressLookupTable(this);
 	PaletteFlag = SupportsPalettes();
 }
@@ -617,6 +622,9 @@ HRESULT STDMETHODCALLTYPE Direct3DDevice8::BeginScene()
 }
 HRESULT STDMETHODCALLTYPE Direct3DDevice8::EndScene()
 {
+	if (App::Instance().EndSceneCallback)
+		App::Instance().EndSceneCallback();
+
 	return ProxyInterface->EndScene();
 }
 HRESULT STDMETHODCALLTYPE Direct3DDevice8::Clear(DWORD Count, const D3DRECT *pRects, DWORD Flags, D3DCOLOR Color, float Z, DWORD Stencil)
diff --git a/source/d3d8to9_index_buffer.cpp b/source/d3d8to9_index_buffer.cpp
index 35dff26..c705d7b 100644
--- a/source/d3d8to9_index_buffer.cpp
+++ b/source/d3d8to9_index_buffer.cpp
@@ -2,7 +2,7 @@
  * Copyright (C) 2015 Patrick Mours. All rights reserved.
  * License: https://github.com/crosire/d3d8to9#license
  */
-
+#include "pch.h"
 #include "d3d8to9.hpp"
 
 Direct3DIndexBuffer8::Direct3DIndexBuffer8(Direct3DDevice8 *Device, IDirect3DIndexBuffer9 *ProxyInterface) :
diff --git a/source/d3d8to9_surface.cpp b/source/d3d8to9_surface.cpp
index b325be7..555ea80 100644
--- a/source/d3d8to9_surface.cpp
+++ b/source/d3d8to9_surface.cpp
@@ -2,7 +2,7 @@
  * Copyright (C) 2015 Patrick Mours. All rights reserved.
  * License: https://github.com/crosire/d3d8to9#license
  */
-
+#include "pch.h"
 #include "d3d8to9.hpp"
 
 Direct3DSurface8::Direct3DSurface8(Direct3DDevice8 *Device, IDirect3DSurface9 *ProxyInterface) :
diff --git a/source/d3d8to9_swap_chain.cpp b/source/d3d8to9_swap_chain.cpp
index 07cf718..8e6f764 100644
--- a/source/d3d8to9_swap_chain.cpp
+++ b/source/d3d8to9_swap_chain.cpp
@@ -2,7 +2,7 @@
  * Copyright (C) 2015 Patrick Mours. All rights reserved.
  * License: https://github.com/crosire/d3d8to9#license
  */
-
+#include "pch.h"
 #include "d3d8to9.hpp"
 
 Direct3DSwapChain8::Direct3DSwapChain8(Direct3DDevice8 *Device, IDirect3DSwapChain9 *ProxyInterface) :
diff --git a/source/d3d8to9_texture.cpp b/source/d3d8to9_texture.cpp
index 6b87e59..4868e81 100644
--- a/source/d3d8to9_texture.cpp
+++ b/source/d3d8to9_texture.cpp
@@ -2,7 +2,7 @@
  * Copyright (C) 2015 Patrick Mours. All rights reserved.
  * License: https://github.com/crosire/d3d8to9#license
  */
-
+#include "pch.h"
 #include "d3d8to9.hpp"
 
 Direct3DTexture8::Direct3DTexture8(Direct3DDevice8 *Device, IDirect3DTexture9 *ProxyInterface) :
diff --git a/source/d3d8to9_vertex_buffer.cpp b/source/d3d8to9_vertex_buffer.cpp
index 13711ea..8a9b181 100644
--- a/source/d3d8to9_vertex_buffer.cpp
+++ b/source/d3d8to9_vertex_buffer.cpp
@@ -2,7 +2,7 @@
  * Copyright (C) 2015 Patrick Mours. All rights reserved.
  * License: https://github.com/crosire/d3d8to9#license
  */
-
+#include "pch.h"
 #include "d3d8to9.hpp"
 
 Direct3DVertexBuffer8::Direct3DVertexBuffer8(Direct3DDevice8 *Device, IDirect3DVertexBuffer9 *ProxyInterface) :
diff --git a/source/d3d8to9_volume.cpp b/source/d3d8to9_volume.cpp
index 3bb9428..a4d0ed2 100644
--- a/source/d3d8to9_volume.cpp
+++ b/source/d3d8to9_volume.cpp
@@ -2,7 +2,7 @@
  * Copyright (C) 2015 Patrick Mours. All rights reserved.
  * License: https://github.com/crosire/d3d8to9#license
  */
-
+#include "pch.h"
 #include "d3d8to9.hpp"
 
 Direct3DVolume8::Direct3DVolume8(Direct3DDevice8 *Device, IDirect3DVolume9 *ProxyInterface) :
diff --git a/source/d3d8types.cpp b/source/d3d8types.cpp
index b190243..5d7981e 100644
--- a/source/d3d8types.cpp
+++ b/source/d3d8types.cpp
@@ -2,7 +2,7 @@
  * Copyright (C) 2015 Patrick Mours. All rights reserved.
  * License: https://github.com/crosire/d3d8to9#license
  */
-
+//#include "pch.h"
 #include "d3d8types.hpp"
 #include <assert.h>
 #include <algorithm>
diff --git a/source/interface_query.cpp b/source/interface_query.cpp
index 938c20b..de7d6d7 100644
--- a/source/interface_query.cpp
+++ b/source/interface_query.cpp
@@ -2,7 +2,7 @@
  * Copyright (C) 2015 Patrick Mours. All rights reserved.
  * License: https://github.com/crosire/d3d8to9#license
  */
-
+//#include "pch.h"
 #define INITGUID
 #include "d3d8to9.hpp"
 
